gson/src/main/java/com/google/gson/internal/NonNullElementWrapperList.java

```diff
@@ -0,0 +1,98 @@
+package com.google.gson.internal;
+
+import java.util.AbstractList;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.Objects;
+import java.util.RandomAccess;
+
+/**
+ * {@link List} which wraps another {@code List} but prevents insertion of
+ * {@code null} elements. Methods which only perform checks with the element
+ * argument (e.g. {@link #contains(Object)}) do not throw exceptions for
+ * {@code null} arguments.
+ */
+public class NonNullElementWrapperList<E> extends AbstractList<E> implements RandomAccess {
+  // Explicitly specify ArrayList as type to guarantee that delegate implements RandomAccess
+  private final ArrayList<E> delegate;
+
+  public NonNullElementWrapperList(ArrayList<E> delegate) {
+    this.delegate = Objects.requireNonNull(delegate);
+  }
+
+  @Override public E get(int index) {
+    return delegate.get(index);
+  }
+
+  @Override public int size() {
+    return delegate.size();
+  }
+
+  private E nonNull(E element) {
+    if (element == null) {
+      throw new NullPointerException("Element must be non-null");
+    }
+    return element;
+  }
+
+  @Override public E set(int index, E element) {
+    return delegate.set(index, nonNull(element));
+  }
+
+  @Override public void add(int index, E element) {
+    delegate.add(index, nonNull(element));
+  }
+
+  @Override public E remove(int index) {
+    return delegate.remove(index);
+  }
+
+  /* The following methods are overridden because their default implementation is inefficient */
+
+  @Override public void clear() {
+    delegate.clear();
+  }
+
+  @Override public boolean remove(Object o) {
+    return delegate.remove(o);
+  }
+
+  @Override public boolean removeAll(Collection<?> c) {
+    return delegate.removeAll(c);
+  }
+
+  @Override public boolean retainAll(Collection<?> c) {
+    return delegate.retainAll(c);
+  }
+
+  @Override public boolean contains(Object o) {
+    return delegate.contains(o);
+  }
+
+  @Override public int indexOf(Object o) {
+    return delegate.indexOf(o);
+  }
+
+  @Override public int lastIndexOf(Object o) {
+    return delegate.lastIndexOf(o);
+  }
+
+  @Override public Object[] toArray() {
+    return delegate.toArray();
+  }
+
+  @Override public <T> T[] toArray(T[] a) {
+    return delegate.toArray(a);
+  }
+
+  @Override public boolean equals(Object o) {
+    return delegate.equals(o);
+  }
+
+  @Override public int hashCode() {
+    return delegate.hashCode();
+  }
+
+  // TODO: Once Gson targets Java 8 also override List.sort
```

Comment:
I think you could override it now, couldn't you? Just don't use the `@Override` annotation. Maybe that will trigger annoying warnings, though.

Reply:
Good point, I did not think of this. Unfortunately `ArrayList.sort` does not exist in Java 7 so it would not provide a great advantage to override this now.

The reason why I added this comment is that `ArrayList.sort` directly [sorts the internal array](https://github.com/openjdk/jdk/blob/d3781ac8a38943d8a20304e770b01d5418ee33d0/src/java.base/share/classes/java/util/ArrayList.java#L1719-L1721), whereas the default `List.sort` [creates a new array and re-orders the elements through the iterator](https://github.com/openjdk/jdk/blob/d3781ac8a38943d8a20304e770b01d5418ee33d0/src/java.base/share/classes/java/util/List.java#L508-L516), so it would be nice to delegate to the more efficient `ArrayList` implementation.

Either way, this is only a nice-to-have feature but it is not required for correctness.

Reply:
Yes, I was thinking you could call `Collections.sort` but that actually just delegates back to `List.sort`. So there isn't really anything you can usefully do without requiring Java 8.